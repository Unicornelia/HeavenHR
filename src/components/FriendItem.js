import React from "react";
import classnames from "classnames";

import TableCell from "@material-ui/core/TableCell";
import TableRow from "@material-ui/core/TableRow";
import DeleteIcon from "@material-ui/icons/Delete";
import IconButton from "@material-ui/core/IconButton";

const FriendItem = ({ id, name, gender, isStarred }) => {

  return (
    <TableRow key={id}>
      <TableCell className="text-secondary"> {id} </TableCell>
      <TableCell className="text-secondary"> {name} </TableCell>
      <TableCell>
        <i
          className={classnames(`fa`, "text-primary", {
            "fa-female": gender === "female",
            "fa-male": gender === "male"
          })}
        />
      </TableCell>
      <TableCell>
        {
          <button type="button" className="btn btn-light text-primary">
            <i
              className={classnames("fa", {
                "fa-star": isStarred,
                "fa-star-o": !isStarred
              })}
            />
          </button>
        }
      </TableCell>
      <TableCell>
        <IconButton area-label="delete" onClick={() => console.log("delete")}>
          <DeleteIcon />
        </IconButton>
      </TableCell>
    </TableRow>
  );
};

export default FriendItem;
