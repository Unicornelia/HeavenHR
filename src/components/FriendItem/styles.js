import { makeStyles } from "@material-ui/core";

export const useStyles = makeStyles(theme => ({
  paper: {
    position: "absolute",
    width: 300,
    backgroundColor: theme.palette.background.paper,
    border: "2px solid #000",
    boxShadow: theme.shadows[5],
    padding: theme.spacing(2, 4, 3),
    textAlign: "center"
  },
  buttons: {
    display: "flex",
    marginTop: theme.spacing(2),
    justifyContent: "space-evenly"
  }
}));

export function getModalStyle() {
  const top = 50;
  const left = 50;

  return {
    top: `${top}%`,
    left: `${left}%`,
    transform: `translate(-${top}%, -${left}%)`
  };
}
