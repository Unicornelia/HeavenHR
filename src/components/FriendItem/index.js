import React, { useState } from "react";
import classnames from "classnames";
import { connect } from "react-redux";

import {
  TableCell,
  TableRow,
  IconButton,
  Tooltip,
  Modal,
  TextField,
  Button
} from "@material-ui/core";
import BorderColorIcon from "@material-ui/icons/BorderColor";

import { updateFriend } from "../../actions";
import { getModalStyle, useStyles } from "./styles";

const FriendItem = props => {
  const { id, name, gender, isStarred } = props;
  const classes = useStyles();
  const [itemStar, setItemStar] = useState(isStarred);
  const [isEditing, setIsEditing] = useState(false);
  const [itemName, setItemName] = useState(name);

  const [modalStyle] = React.useState(getModalStyle);

  //TODO needs fixing
  const toggleItemStar = () => {
    setItemStar(!itemStar);
  };

  const handleStarClick = () => {
    toggleItemStar();
    const updatedFriend = { id, name, gender, isStarred: itemStar };
    props.onUpdateFriend(updatedFriend);
  };

  const handleUpdateName = e => {
    setItemName(e.target.value);
  };

  const handleSubmitUpdatedName = () => {
    const updatedFriend = { id, name: itemName, gender, isStarred };
    props.onUpdateFriend(updatedFriend);
    setIsEditing(false);
  };

  const handleClose = () => {
    setIsEditing(false);
  };

  return (
    <TableRow key={id} className="tableRow">
      <TableCell className="text-secondary">{id}</TableCell>
      <TableCell className="text-secondary">
        {name}
        <Tooltip title={"Update name"}>
          <IconButton
            className="iconButton"
            area-label="update"
            fontSize="small"
            onClick={() => setIsEditing(true)}
          >
            <BorderColorIcon />
          </IconButton>
        </Tooltip>
        {isEditing ? (
          <Modal
            open={isEditing}
            onClose={handleClose}
            aria-labelledby="simple-modal-title"
            aria-describedby="simple-modal-description"
          >
            <div style={modalStyle} className={classes.paper}>
              <TextField
                id="standard-disabled"
                label="name"
                defaultValue={itemName}
                onChange={e => handleUpdateName(e)}
              />
              <span className={classes.buttons}>
                <Button
                  id="submit"
                  variant="outlined"
                  color="secondary"
                  onClick={e => handleSubmitUpdatedName(e)}
                >
                  Submit
                </Button>
                <Button
                  id="cancel"
                  variant="outlined"
                  color="secondary"
                  onClick={handleClose}
                >
                  Cancel
                </Button>
              </span>
            </div>
          </Modal>
        ) : null}
      </TableCell>
      <TableCell>
        <i
          className={classnames(`fa`, "text-primary", {
            "fa-female": gender === "female",
            "fa-male": gender === "male"
          })}
        />
      </TableCell>
      <TableCell>
        {
          <Button
            id="star"
            type="button"
            className="btn btn-light text-primary"
            onClick={handleStarClick}
          >
            {itemStar ? (
              <i className={"fa fa-star"} />
            ) : (
              <i className={"fa fa-star-o"} />
            )}
          </Button>
        }
      </TableCell>
    </TableRow>
  );
};

const mapStateToProps = friends => ({
  friends: friends
});

const mapDispatchToProps = dispatch => {
  return {
    onUpdateFriend: friend => {
      dispatch(updateFriend(friend));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(FriendItem);
