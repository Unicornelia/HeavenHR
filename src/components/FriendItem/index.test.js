import React from "react";
import configureStore from "redux-mock-store";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure, mount } from "enzyme";
import { Provider } from "react-redux";
import faker from "faker";

import FriendItem from "./index";
configure({ adapter: new Adapter() });

describe("<FriendItem/>", () => {
  const mockStore = configureStore();
  const initialState = [
    {
      id: faker.random.uuid(),
      name: faker.name.findName(),
      gender: "male",
      isStarred: false
    },
    {
      id: faker.random.uuid(),
      name: faker.name.findName(),
      gender: "female",
      isStarred: false
    }
  ];
  const store = mockStore(initialState);

  it("should match snapshot", () => {
    const wrapper = shallow(
      <Provider store={store}>
        <FriendItem />
      </Provider>
    );
    const component = wrapper.dive();

    expect(component).toMatchSnapshot();
    expect(store.getActions()).toMatchSnapshot();
  });

  it("should render the connected friend item", () => {
    const friend = {
      id: faker.random.uuid(),
      name: faker.name.findName(),
      gender: "female",
      isStarred: false
    };
    const wrapper = mount(
      <Provider store={store}>
        <FriendItem friend={friend} />
      </Provider>
    );

    expect(wrapper.find("FriendItem")).toBeDefined();
  });

  it("should find star button", () => {
    const friend = {
      id: "some id",
      name: "Some Name",
      gender: "male",
      isStarred: true
    };
    const wrapper = mount(
      <Provider store={store}>
        <FriendItem friend={friend} />
      </Provider>
    );

    expect(wrapper.find("button#star")).toHaveLength(1);
  });
});
