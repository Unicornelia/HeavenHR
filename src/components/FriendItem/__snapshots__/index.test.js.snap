// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FriendItem/> should match snapshot 1`] = `ShallowWrapper {}`;

exports[`<FriendItem/> should match snapshot 2`] = `Array []`;
